video/out/vo_drm.c

```diff
@@ -0,0 +1,523 @@
+    dev->conn = conn->connector_id;
+    dev->front_buf = 0;
+    memcpy(&dev->mode, &conn->modes[0], sizeof(dev->mode));
+    dev->bufs[0].width = conn->modes[0].hdisplay;
+    dev->bufs[0].height = conn->modes[0].vdisplay;
+    dev->bufs[1].width = conn->modes[0].hdisplay;
+    dev->bufs[1].height = conn->modes[0].vdisplay;
+
+    MP_INFO(vo, "Connector using mode %ux%u\n",
+            dev->bufs[0].width, dev->bufs[0].height);
+
+    ret = modeset_find_crtc(vo, fd, res, conn, dev);
+    if (ret) {
+        MP_ERR(vo, "Connector %d has no valid CRTC\n", conn_id);
+        goto end;
+    }
+
+    unsigned int i, j;
+    for (i = 0; i < BUF_COUNT; i ++) {
+        ret = modeset_create_fb(vo, fd, &dev->bufs[i]);
+        if (ret) {
+            for (j = 0; j < i; j ++) {
+                modeset_destroy_fb(fd, &dev->bufs[j]);
+            }
+            MP_ERR(vo, "Cannot create framebuffer for connector %d\n",
+                   conn_id);
+            return ret;
+        }
+    }
+
+end:
+    if (conn) {
+        drmModeFreeConnector(conn);
+        conn = NULL;
+    }
+    if (res) {
+        drmModeFreeResources(res);
+        res = NULL;
+    }
+    if (ret == 0) {
+        *out = dev;
+    } else {
+        talloc_free(dev);
+    }
+    return ret;
+}
+
+
+
+static int reconfig(struct vo *vo, struct mp_image_params *params, int flags)
+{
+    struct priv *p = vo->priv;
+    float device_ar = p->device_w / (float)p->device_h;
+
+    vo_get_src_dst_rects(vo, &p->src, &p->dst, &p->osd);
+    int32_t video_w = (p->dst.x1 - p->dst.x0);
+    int32_t video_h = (p->dst.y1 - p->dst.y0);
+    float video_ar = video_w / (float)video_h;
+    int32_t w, h;
+    if (device_ar > video_ar) {
+        w = p->device_h * video_ar;
+        h = p->device_h;
+    } else {
+        w = p->device_w;
+        h = p->device_w * video_ar;
+    }
+    if (w > p->device_w) w = p->device_w;
+    if (h > p->device_h) h = p->device_h;
+
+    // p->osd contains the parameters assuming OSD rendering in window
+    // coordinates, but OSD can only be rendered in the intersection
+    // between window and video rectangle (i.e. not into panscan borders).
+    p->osd.w = w;
+    p->osd.h = h;
+    p->osd.mt = MPMIN(0, p->osd.mt);
+    p->osd.mb = MPMIN(0, p->osd.mb);
+    p->osd.mr = MPMIN(0, p->osd.mr);
+    p->osd.ml = MPMIN(0, p->osd.ml);
+
+    p->x = (p->device_w - w) >> 1;
+    p->y = (p->device_h - h) >> 1;
+    p->buf = (char*)talloc_size(vo, p->device_w * p->device_h * 4);
+    memset(p->buf, 0, p->device_w * p->device_h * 4);
+
+    mp_sws_set_from_cmdline(p->sws, vo->opts->sws_opts);
+    p->sws->src = *params;
+    p->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT_BGR0,
+        .w = w,
+        .h = h,
+        .d_w = w,
+        .d_h = h,
+    };
+
+    MP_INFO(vo, "resizing %dx%d -> %dx%d\n",
+            p->sws->src.w,
+            p->sws->src.h,
+            p->sws->dst.w,
+            p->sws->dst.h);
```

Comment:
Kind of pointless; similar and more comprehensive info is printed in verbose mode by common code.